File: SRM 681/FleetFunding.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashSet;
import java.util.TreeSet;

public class FleetFunding {
	public int maxShips(int m, int[] k, int[] a, int[] b) {
		Factory[] factories = new Factory[k.length];
		int total = 0;
		for (int i = 0; i < k.length; ++i) {
			factories[i] = new Factory(i, a[i], b[i], k[i]);
			total += k[i];
		}
		Arrays.sort(factories, new Comparator<Factory>() {
			@Override
			public int compare(Factory o1, Factory o2) {
				return o1.a - o2.a;
			}
		});
		int left = 0;
		int right = total / m + 1;
		outer:
		while (right - left > 1) {
			int need = (left + right) / 2;
			int ptr = 0;
			TreeSet<Factory> available = new TreeSet<Factory>(new Comparator<Factory>() {
				@Override
				public int compare(Factory o1, Factory o2) {
					if (o1.b != o2.b) return o1.b - o2.b;
					return o1.id - o2.id;
				}
			});

			int at = 1;
			while (at <= m) {
				while (ptr < factories.length && factories[ptr].a <= at) {
					Factory nxt = factories[ptr];
					nxt.remains = nxt.k;
					available.add(nxt);
					++ptr;
				}
				while (!available.isEmpty() && available.first().b < at) {
					available.pollFirst();
				}
				int upto = m;
				if (ptr < factories.length) upto = Math.min(upto, factories[ptr].a - 1);
				if (!available.isEmpty()) upto = Math.min(upto, available.first().b);
				int required = (upto - at + 1) * need;
				while (required > 0) {
					if (available.isEmpty()) {
						right = need;
						continue outer;
					}
					Factory first = available.first();
					int by = Math.min(required, first.remains);
					first.remains -= by;
					required -= by;
					if (first.remains == 0) {
						available.pollFirst();
					}
				}
				at = upto + 1;
			}
			left = need;
		}
		return left;
	}
	static class Factory {
		int id;
		int a;
		int b;
		int k;
		int remains;
		public Factory(int id, int a, int b, int k) {
			this.id = id;
			this.a = a;
			this.b = b;
			this.k = k;
		}
	}

// CUT begin
	public static void main(String[] args){
		System.err.println("FleetFunding (300 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            int m = Integer.parseInt(Reader.nextLine());
            int[] k = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < k.length; ++i)
                k[i] = Integer.parseInt(Reader.nextLine());
            int[] a = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < a.length; ++i)
                a[i] = Integer.parseInt(Reader.nextLine());
            int[] b = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < b.length; ++i)
                b[i] = Integer.parseInt(Reader.nextLine());
            Reader.nextLine();
            int __answer = Integer.parseInt(Reader.nextLine());

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(m, k, a, b, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1471513595;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 300 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(int m, int[] k, int[] a, int[] b, int __expected) {
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		FleetFunding instance = new FleetFunding();
		int __result = 0;
		try {
			__result = instance.maxShips(m, k, a, b);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__result == __expected) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + __expected);
			System.err.println("           Received: " + __result);
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "/Users/resry/topcoder/SRM 681/FleetFunding.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}

File: SRM 681/FleetFunding.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>

using namespace std;
#define PB push_back
#define FOR(i, a, b) for(int i =(a); i <=(b); ++i)
#define RE(i, n) FOR(i, 1, n)
#define SZ(x) (x).size()
#define st first
#define nd second
typedef pair<int, int> PII;
class FleetFunding {
    public:
    int maxShips(int m, vector<int> k, vector<int> a, vector<int> b) {
        
    
    int kl = 0, kp = 1e9, faj = 0;
    vector<vector<int> > wh(m + 2);
    for (int i = 0; i < SZ(a); i++) {
      wh[a[i]].PB(i);
    }
    while (kl <= kp) {
      int aktc = (kl + kp) / 2;
      multiset<PII> secik;
      
      int fail = 0;
      RE (i, m) {
        for (auto p : wh[i]) {
          secik.insert(make_pair(b[p], k[p]));
        }
        int need = aktc;
        while (need > 0) {
          if (secik.empty()) {
            fail = 1;
            break;
          }
          auto p = *secik.begin();
          secik.erase(secik.begin());
          if (p.st < i) {
            continue;
          }
          int got = min(need, p.nd);
          p.nd -= got;
          need -= got;
          if (p.nd) {
            secik.insert(make_pair(p.st, p.nd));
          }
        }
      }
      if (!fail) {
        faj = aktc;
        kl = aktc + 1;
      } else {
        kp = aktc - 1;
      }
    }
    return faj;
    }
};

// CUT begin
ifstream data("FleetFunding.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T> void from_stream(vector<T> &ts) {
    int len;
    from_stream(len);
    ts.clear();
    for (int i = 0; i < len; ++i) {
        T t;
        from_stream(t);
        ts.push_back(t);
    }
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(int m, vector<int> k, vector<int> a, vector<int> b, int __expected) {
    time_t startClock = clock();
    FleetFunding *instance = new FleetFunding();
    int __result = instance->maxShips(m, k, a, b);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        int m;
        from_stream(m);
        vector<int> k;
        from_stream(k);
        vector<int> a;
        from_stream(a);
        vector<int> b;
        from_stream(b);
        next_line();
        int __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(m, k, a, b, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1473118069;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 300 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "FleetFunding (300 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end
